ServicePackaging.md

```diff
@@ -0,0 +1,533 @@
+
+### Edit the Dockerfile and Entrypoint
+
+Next, it's time to edit the `Dockerfile`.  This defines how to build the image for the package by declaring the environment, building stages, and mounting the package to the volume specified in the `Manifest`.
+
+1. We start by importing a base image, in this case Alpine, as recommended.
+
+`FROM alpine:3.13`
+
+2. Next we issue some commands, which add a repository, update, install some required software, and finally create a directory for the nginx software.
+
+```
+RUN echo https://dl-cdn.alpinelinux.org/alpine/edge/community >> /etc/apk/repositories
+RUN apk update
+RUN apk add tini
+RUN apk add bash nginx yq
+
+RUN mkdir /run/nginx
+```
+
+3. Next we will add the `docker_entrypoint.sh` file and `www` directory from the repository as well as copying over some `.css` files for styling.  Then we set the permissions for `docker_entrypoint.sh`.
+
+```
+ADD ./docker_entrypoint.sh /usr/local/bin/docker_entrypoint.sh
+ADD www /var/www
+RUN cp /var/www/assets/main.css /var/www/fuck-off/main.css
+RUN cp /var/www/assets/main.css /var/www/index/main.css
+RUN cp /var/www/assets/main.css /var/www/welcome/main.css
+RUN chmod a+x /usr/local/bin/docker_entrypoint.sh
+```
+
+4. Next we set a working directory, expose a port, and set the location of the entrypoint.
+
+```
+WORKDIR /root
+
+EXPOSE 80
+
+ENTRYPOINT ["/usr/local/bin/docker_entrypoint.sh"]
+```
+
+5. Great, let's take a look at our final Embassy Pages `Dockerfile`:
+
+```
+FROM alpine:3.13
+
+RUN echo https://dl-cdn.alpinelinux.org/alpine/edge/community >> /etc/apk/repositories
+RUN apk update
+RUN apk add tini
+RUN apk add bash nginx yq
+
+RUN mkdir /run/nginx
+
+ADD ./docker_entrypoint.sh /usr/local/bin/docker_entrypoint.sh
+ADD www /var/www
+RUN cp /var/www/assets/main.css /var/www/fuck-off/main.css
+RUN cp /var/www/assets/main.css /var/www/index/main.css
+RUN cp /var/www/assets/main.css /var/www/welcome/main.css
+RUN chmod a+x /usr/local/bin/docker_entrypoint.sh
+
+WORKDIR /root
+
+EXPOSE 80
+
+ENTRYPOINT ["/usr/local/bin/docker_entrypoint.sh"]
+```
+
+6. Okay, let's move on to our `docker_entrypoint.sh` file.  This is a bash script that defines what to do when the service starts.  It will need to complete any environment setup (such as folder substructure), sets any environment variables, and executes the run command.  Let's take a look at our Pages example:
+
+```
+#!/bin/bash
+
+export HOST_IP=$(ip -4 route list match 0/0 | awk '{print $3}')
+
+echo start9/public > .backupignore
+echo start9/shared >> .backupignore
+
+home_type=$(yq e '.homepage.type' start9/config.yaml)
+subdomains=($(yq e '.subdomains.[].name' start9/config.yaml))
+```
+
+7. First we've defined the file as a bash, exported the IP address, chosen folders to ignore during backup, and set the homepage and subdomains types from our config.
+
+***I'VE ABANDONED THIS SECTION AFTER DISCUSSION LEADING TO THE IDEA THAT THIS MAY NOT BE A GOOD EXAMPLE FILE***
```

Comment:
🤣

Reply:
Perhaps `cups` would be better?

Reply:
On second thought we should probably just update the `hello-world` wrapper with all of the features it's missing, like dependencies

Reply:
Leaning toward hello-world updates being the best course of action